feat(graphql): added Error Handler that will throw GraphQL errors

sudeepjd opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://refine.dev/docs/guides-concepts/contributing/#commit-convention

## Bugs / Features

- [x] Related issue(s) linked
- [x] Tests for the changes have been added
- [x] Docs have been added / updated
- [x] Changesets have been added https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset

## What is the current behavior?
Refine's GraphQL package is loosely based on nestjs-query. However, uses urql under the hood. urql uses 2 different types of error handling mechanisms, an errorExchange in exchanges or response.error on the result. The way refine is built using Tanstack query it is expecting an exception new Error to be thrown in order to be caught and treated as an error. But urql does not do this out of the box, and so refine always treats any error as a success, which impact the notifications provider.

## What is the new behavior?
Once the error is received by client from the query or mutation, it passes the **response.error** to an error handler function that parses the error and throws it as Tanstack Query is expecting so that it is treated as an error and not as a success.

Also added in a new *data-provider-graphql* example to use the graphql data provider.

fixes (https://github.com/refinedev/refine/issues/6493): [BUG] @refinedev/graphql Error Handling does not work as it should


## Notes for reviewers

* I have added in an error differentiator prefix. [Code], [Network], [GraphQL] so the error can be classified and handled accordingly, this may be a breaking change for anyone on the [Code] who has tightly coupled with the "Operation is required." is required string.
* Tests have been updated to validate the changes made.
* Documentation has been updated to add in sections for Handling errors based on the prefixes, managing retries and the Authentication has been updated as it was using the nestjs-query mechanism for a custom fetch instead of using the fetchOptions in urql.
* Thank you for allowing me to make this fix,

sudeepjd commented:
Dear Refine Team,

Please check this draft PR and if ok, will submit it through. Thanks.

BatuhanW commented:
Hey @sudeepjd thanks for the PR! I'll make some time to review it tomorrow.

BatuhanW reviewed (commented):
Thanks for the PR! LGTM. You can also add the example you previously added, only thing you need to do is to run `pnpm install` in the root folder, then it will add your example and its dependencies to pnpm-lock.yaml.

sudeepjd commented:
> Thanks for the PR! LGTM. You can also add the example you previously added, only thing you need to do is to run `pnpm install` in the root folder, then it will add your example and its dependencies to pnpm-lock.yaml.

I have now added back in the example/data-provider-graphql and have also updated the documentation on the GraphQL package to point to the new example.

BatuhanW commented:
> > Thanks for the PR! LGTM. You can also add the example you previously added, only thing you need to do is to run `pnpm install` in the root folder, then it will add your example and its dependencies to pnpm-lock.yaml.
> 
> I have now added back in the example/data-provider-graphql and have also updated the documentation on the GraphQL package to point to the new example.
> 
> However, there are a few edits to the **pnpm-lock.yaml** file apart from what would normally be added in for the data-provider-graphql example. I am not sure if those are intended to be changed or not.

Those edits are not problem, although annoying. No worries.

BatuhanW reviewed (approved):
Thanks for the great work 🚀
